kvclient: optimize and clean up sorting computation

andrewbaptist opened the pull request:
Previously the locality distance and the latency function were computed multiple times for each node in the sort.Slice method. This change computes the values once when the ReplicaSlice is created and uses simple comparisions within the sorting loop.

Epic: none
Informs: #112351

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113942)
<!-- Reviewable:end -->

Commit:
kvclient: optimize and clean up sorting computation

Previously the locality distance, latency and healthfunctions were
computed multiple times for each node in the sort.Slice method.
This change computes the values once when the ReplicaSlice is
created and uses simple comparisions within the sorting loop.

Note that the added fields are only used in the Optimize loop.

Epic: none
Informs: #112351

Release note: None

arulajmani reviewed (commented):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NityWpd4eZU7s0ipj6h):*
nit: as written, we're computing these values in `OptimizeReplicaOrder`, not when the `ReplicaSlice` is created. We should use local variables instead of saving these fields on `ReplicaInfo` instead.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NitxzpP-omWVrQEoZGv:-NitxzpP-omWVrQEoZGw:blps3n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L198)):*
> ```Go
> 		return
> 	}
> 	// populate the tier match length and locality before starting the loop.
> ```

nit: s/populate/Populate/g
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NitxCUCEOVLg7fqNIfc:-NitxCUD9xpC2uxq1wuW:bqqmswe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L200)):*
> ```Go
> 	// populate the tier match length and locality before starting the loop.
> 	for i := range rs {
> 		rs[i].tierMatchLength = locality.SharedPrefix(rs[i].Locality)
> ```

It's a bit odd that we're modifying `ReplicaInfo.tierMatchLength` and `ReplicaInfo.latency` here, on the receiver, and then reading them right below. Should these be local function variables instead?
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nitxk-r-36LjAQPNA3q:-Nitxk-r-36LjAQPNA3r:b-54k80e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7d1479743f71d7fa9f1215f6d9ba4fdfaad1d22/pkg/kv/kvclient/kvcoord/replica_slice.go#L214)):*
> ```Go
> 	sort.Slice(rs, func(i, j int) bool {
> 		// Replicas on the same node have the same latency.
> 		if rs[i].NodeID == rs[j].NodeID {
> ```

Did you mean to remove this because this isn't possible? 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nityuh05ipxt1P9zfnp:-Nityuh05ipxt1P9zfnq:b-7x047y) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L29)):*
> ```Go
> 	roachpb.ReplicaDescriptor
> 	Locality        roachpb.Locality
> 	tierMatchLength int
> ```

Consider adding a `// Fields precomputed when ordering replicas in a range.` above this line.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NityXUbDkAcf_weZjzQ:-NityXUbDkAcf_weZjzR:b-ctok4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L198)):*
> ```Go
> 		return
> 	}
> 	// populate the tier match length and locality before starting the loop.
> ```

nit: "Populate"
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NitxCUCEOVLg7fqNIfc:-Nitz7pB6msYicLJ4VIQ:b-emk2ms) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L200)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

It's a bit odd that we're modifying `ReplicaInfo.tierMatchLength` and `ReplicaInfo.latency` here, on the receiver, and then reading them right below. Should these be local function variables instead?
</blockquote></details>

I think the point is to avoid a new heap allocation by hanging this information on the existing slice.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nitxk-r-36LjAQPNA3q:-NitzH5EDKp3b3uDZMtJ:bpjx0jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7d1479743f71d7fa9f1215f6d9ba4fdfaad1d22/pkg/kv/kvclient/kvcoord/replica_slice.go#L214)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did you mean to remove this because this isn't possible?
</blockquote></details>

I'm also not sure. Why remove this? If two replicas are on the same node then the following checks should probably all fall through and we should return false, but that's not actually the case if `rs[i].NodeID == nodeID`.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NityWpd4eZU7s0ipj6h):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: as written, we're computing these values in `OptimizeReplicaOrder`, not when the `ReplicaSlice` is created. We should use local variables instead of saving these fields on `ReplicaInfo` instead.
</blockquote></details>

As we're not going with the local variables approach to avoid heap allocations, let's update the commit message to better reflect where these fields are assigned. 
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nityuh05ipxt1P9zfnp:-Niuu44sApt3fbdLl4uh:b-1ww0mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding a `// Fields precomputed when ordering replicas in a range.` above this line.
</blockquote></details>

Could you also add some words about who the latency and tierMatchLength is between?
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NitxCUCEOVLg7fqNIfc:-NiutlqD4AZ047R1NRab:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L200)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think the point is to avoid a new heap allocation by hanging this information on the existing slice.
</blockquote></details>

Ack.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NityWpd4eZU7s0ipj6h):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

As we're not going with the local variables approach to avoid heap allocations, let's update the commit message to better reflect where these fields are assigned.
</blockquote></details>

I added this to both the commit and a header on the method. I have another PR which makes these structs more "private" that I'll put up later today as part of #114033. But it is easier to pull this one out.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nityuh05ipxt1P9zfnp:-Nj84FYfDCMMvlqoEJ2-:b-hck344) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L29)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you also add some words about who the latency and tierMatchLength is between?
</blockquote></details>

Added text.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NityXUbDkAcf_weZjzQ:-Nj84N1b7YxRnH-wOBd6:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L198)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: "Populate"
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-NitxCUCEOVLg7fqNIfc:-Nj84PGdBdL6nI3k4UMc:bmdzy1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9a90d314d51d5c28bd4bc692a56dcb7e212134/pkg/kv/kvclient/kvcoord/replica_slice.go#L200)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Ack.
</blockquote></details>

Yep - this is not great right now, but I'm hoping to clean this up more in the next few commits.
___
*[`pkg/kv/kvclient/kvcoord/replica_slice.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-Nitxk-r-36LjAQPNA3q:-Nj84YzG5y2MqAcw_Qz5:b-a4c950) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7d1479743f71d7fa9f1215f6d9ba4fdfaad1d22/pkg/kv/kvclient/kvcoord/replica_slice.go#L214)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm also not sure. Why remove this? If two replicas are on the same node then the following checks should probably all fall through and we should return false, but that's not actually the case if `rs[i].NodeID == nodeID`.
</blockquote></details>

You're both correct that it was wrong before. I simplified the sorting a lot by introducing a -1 value for local node latency. This handles all the cases the same as before and I think  is much simpler to reason about since there are less special cases. The key to note is that if NodeID is equal, then it will always drop to the last comparison and now always return false. We don't need to special case with checking if we are "on the local node" anymore.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113942#-:-Nj90wFf2waEJCy-6A7r:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)



<!-- Sent from Reviewable.io -->

andrewbaptist commented:
bors r=nvanbenschoten

TFTR!
